### PR TITLE
txn: Split the read and write part of the ResolveLock command.

### PR DESCRIPTION
**Pull request by longfangsong**
Signed-off-by: longfangsong <longfangsong@icloud.com>

### What problem does this PR solve?

Problem Summary:

The `ResolveLock` used to do both read and write operations, which is weird.

### What is changed and how it works?

What's Changed:

Spilt `ResolveLock` into `ScanStaleLocks`(the read part) and `ResolveLock`(the write part)

### Check List 

Tests 

- Unit test

### CODE REVIEW

File: src/storage/txn/commands/scan_stale_locks.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+use tikv_util::collections::HashMap;
+use txn_types::{Key, TimeStamp};
+
+command! {
+    /// Scan locks for resolving according to `txn_status`.
+    ///
+    /// During the GC operation, this should be called to find out stale locks whose timestamp is
+    /// before safe point.
+    /// This should follow by a `ResolveLock`
+    ScanStaleLocks:
```
Comment (longfangsong) [inline src/storage/txn/commands/scan_stale_locks.rs:13]: Tell me if you had a better name for this.

File: src/storage/txn/commands/scan_stale_locks.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            /// `txn_status`. `"k4"` will not be affected either, because it doesn't have a non-committed
+            /// version.
+            txn_status: HashMap<TimeStamp, TimeStamp>,
+            scan_key: Option<Key>,
+        }
+}
+
+impl CommandExt for ScanStaleLocks {
+    ctx!();
+    tag!(resolve_lock);
```
Comment (longfangsong) [inline src/storage/txn/commands/scan_stale_locks.rs:40]: Maybe we should also add `scan_stale_locks` into `crate::storage::metrics::CommandKind`?
I'm not very sure about how will that effect the other parts of the TiDB-pd-TiKV system.

File: src/storage/txn/commands/scan_stale_locks.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    /// Scan locks for resolving according to `txn_status`.
+    ///
+    /// During the GC operation, this should be called to find out stale locks whose timestamp is
+    /// before safe point.
+    /// This should follow by a `ResolveLock`
+    ScanStaleLocks:
+        cmd_ty => (),
+        display => "kv::scan_stale_locks", (),
+        content => {
+            /// Maps lock_ts to commit_ts. If a transaction was rolled back, it is mapped to 0.
```
Comment (longfangsong) [inline src/storage/txn/commands/scan_stale_locks.rs:17]: These comments about `txn_status` exists both in `ScanStaleLocks` and `ResolveLock`.
I think maybe we can keep them in one of them and reference it on the other?

**Comment by longfangsong** [comment]
@nrc PTAL.

File: src/storage/txn/commands/scan_stale_locks.rs (hunk 1 above)
Comment (sticnarf) [reply src/storage/txn/commands/scan_stale_locks.rs:13]: Sounds good

File: src/storage/txn/commands/scan_stale_locks.rs (hunk 2 above)
Comment (sticnarf) [reply src/storage/txn/commands/scan_stale_locks.rs:40]: I think we can. AFAIK this won't affect scheduling.

File: src/storage/txn/commands/scan_stale_locks.rs (hunk 1 above)
Comment (sticnarf) [reply src/storage/txn/commands/scan_stale_locks.rs:13]: As this command will schedule a `ResolveLock` command, what about calling it `ScanAndResolveLocks`?

File: src/storage/txn/commands/scan_stale_locks.rs (hunk 2 above)
Comment (nrc) [reply src/storage/txn/commands/scan_stale_locks.rs:40]: I think we should do that, but maybe @youjiali1995 has an opinion?

File: src/storage/txn/commands/scan_stale_locks.rs (hunk 3 above)
Comment (nrc) [reply src/storage/txn/commands/scan_stale_locks.rs:17]: > I think maybe we can keep them in one of them and reference it on the other?

Sounds good.

File: src/storage/txn/commands/scan_stale_locks.rs (hunk 1 above)
Comment (nrc) [reply src/storage/txn/commands/scan_stale_locks.rs:13]: ResolveLocksReadPhase?

Comment (longfangsong) [reply src/storage/txn/commands/scan_stale_locks.rs:13]: > As this command will schedule a `ResolveLock` command, what about calling it `ScanAndResolveLocks`?

I've also consider about this. But `ResolveLock` **is not part of this command**.

Comment (longfangsong) [reply src/storage/txn/commands/scan_stale_locks.rs:13]: > ResolveLocksReadPhase?

This one sounds good.

**Pull request merged**